### ISSUE TITLE
1, 5~7 설명 정리

### ISSUE THREAD

**Issue by jpotw** [Basic-Track-Team5/term-project#4]
## 1 : 주제 및 데이터셋 소개

scikit-learn을 이용해 RT-IoT2022 센서 데이터를 기반으로, 침입 여부와 공격 유형을 분류하는 머신러닝 모델을 설계하는 것을 목표로 프로젝트를 진행했습니다. 

데이터셋에 대한 설명:
- IoT(Internet of Things; 사물 인터넷) 환경의 사이버 보안 문제를 해결하기 위해 만들어진 가상의 RT(Real Time) 네트워크 트래픽 데이터
  - IoT 기기들이 인터넷 상에서 통신할 때 주고받는 네트워크 트래픽을 관찰하고 그 특성들을 기록한 데이터
  - '센서 값'(e.g. 센서가 측정한 온도, 습도, 소리) 그 자체가 아니라, 그런 센서 값을 보내거나 명령을 받을 때 발생하는 네트워크 통신 패턴을 분석한 데이터
  - 해당 데이터는 실제 IoT 인프라에서 정상적인 통신과 다양한 해킹 공격 상황을 연출하여 수집되었음 (현재 데이터셋 기준 정상 1: 공격 10 의 불균형 문제가 존재함)

- RT-IoT2022 데이터셋은 실제(에 유사한) IoT 환경의 트래픽을 담고 있기 때문에, 이 데이터셋을 활용하면 IoT 환경에 특화된, 더 효과적이고 정확한 침입 탐지 시스템(IDS)을 개발하고 그 성능을 평가할 수 있음. 그래서 주로 IoT 네트워크 보안과 관련해 가장 대표적으로 사용되는 데이터셋 중 하나.

## 5 : 결과 설명

- 프로젝트에서 RT-IoT2022 데이터셋을 기반으로 이진 분류 및 다중 분류 모델을 구축하고 PCA 적용 전후의 성능을 확인하였음.
- 이진 분류에서는 센서 데이터가 정상적인 신호(normal)인지 공격 신호(attack)인지 여부를 분류하였음
- 다중 분류에서는 공격 유형을 세부적으로 분류하였음.
- 주요 평가지표: Accuracy와 F1-score
  - 데이터 불균형(Normal:Attack $\approx$ 1:10)이 존재하기 때문에 Accuracy 뿐만 아니라 F1-score도 측정하였음

### 전반적인 결론

- 평가지표 기준 이진 분류, 다중 분류 모두 매우 높은 성능을 보였음
- RT-IoT2022 데이터셋은 적용된 전처리(특히 SMOTE)를 통해 트리 기반 앙상블 모델을 사용하여 이진 및 다중 클래스 시나리오 모두에서 매우 정확한 침입 탐지가 가능했음
- XGBoost와 LightGBM (적절히 튜닝되었거나 기본/양호한 기본 파라미터를 사용한 경우)은 최고 수준의 성능을 보여줌
- 30개 구성 요소로의 PCA는 여전히 매우 높은 성능을 유지하였으나 (특히 데이터 수가 적은 클래스에서) 약간의 성능 저하를 초래
- 구체적인 결과 수치는 아래 모델 성능 요약표 참고

**모델 성능 요약표**

---

### **I. 이진 분류 (Attack vs. Normal)**

| 모델              | 데이터 처리                               | 정확도 (Accuracy) | F1-Score | 주요 관찰 및 해석                                                                                                |
| :---------------- | :---------------------------------------- | :---------------- | :--------------------- | :--------------------------------------------------------------------------------------------------------------- |
| **Random Forest (RF)** | No SMOTE (class\_weight='balanced')       | 0.9986            | 0.9933                 | `class_weight`로 불균형 완화, 매우 우수한 성능.                                                                     |
|                   | SMOTE                                     | 0.9987            | 0.9936                 | `class_weight` 사용 시보다 약간 더 우수. SMOTE가 유용한 합성 샘플 생성.                                                  |
|                   | PCA + SMOTE                               | 0.9973            | 0.9872                 | PCA로 성능 약간 저하. PCA가 일부 미묘한 정보 손실 유발 가능성. 30개 주성분으로의 축소가 RF에 최적이 아니었을 수 있음.          |
| **LightGBM (LGBM)** | SMOTE (Hyperopt 튜닝)                     | ~0.9990           | 0.99                   | 매우 우수한 성능.                                                                                                  |
|                   | PCA + SMOTE (Hyperopt 튜닝)               | ~0.9979           | 0.99                   | RF와 유사하게 PCA로 인해 성능 약간 저하 (오탐/미탐 증가, 여전히 매우 적음). |

---

### **II. 다중 분류 (개별 공격 유형 및 Normal)**

| 모델              | 데이터 처리                                  | 정확도 (Accuracy) | F1-Score | 주요 관찰 및 해석                                                                                                                                                              |
| :---------------- | :------------------------------------------- | :---------------- | :----------------- | :----------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **Random Forest (RF)** | No SMOTE (GridSearchCV 튜닝, rf\_clf1)       | 0.9982            | 0.9982             | SMOTE 없이도 매우 우수한 성능                                                                                           |
|                   | PCA + SMOTE (GridSearchCV 튜닝)              | 0.9972            | 0.9972             | PCA로 전반적 성능 약간 저하. 매우 드문 클래스(Metasploit_Brute_Force_SSH F1: 0.83, NMAP_FIN_SCAN F1: 0.80) 분류가 더 어려움.                                                              |
| **LightGBM (LGBM)** | SMOTE (튜닝 안 함, 기본 모델)                  | ~1.00             | ~1.00              | 매우 우수한 성능                                                                                              |
|                   | PCA + SMOTE (Hyperopt 튜닝, `max_evals=10`)  | ~1.00             | ~1.00              | 매우 우수한 성능                                               |
| **XGBoost**       | SMOTE (Hyperopt 튜닝, `max_evals=10`)        | ~1.00             | ~1.00              | 매우 우수한 성능                                                                                                |
|                   | PCA + SMOTE (Hyperopt 튜닝, `max_evals=10`)  | ~1.00             | ~1.00              | 전반적으로 매우 우수하나, XGBoost에서 PCA가 NMAP_FIN_SCAN 식별에 부정적 영향 (F1: 0.57).                                                                                             |

---

## 6: 결과가 높게 나온 이유에 대한 분석

RT-IoT2022 데이터셋을 기반으로 다양한 머신러닝 모델을 실험한 결과, 대부분의 분류 모델이 매우 높은 성능을 기록하였음(ex. 이진 분류에서는 Gradient Boosting, Random Forest 등의 모델이 Accuracy, Precision, Recall, F1 Score 전 항목에서 1.00이라는 정확한 수치를 나타냄) 

[RT-IoT2022 데이터셋을 활용한 Airlangga (2024)의 논문](https://www.researchgate.net/publication/379979592_Comparative_Analysis_of_Machine_Learning_Models_for_Intrusion_Detection_in_Internet_of_Things_Networks_Using_the_RT-IoT2022_Dataset)에서도 우리와 유사하게 머신러닝 모델들이 매우 높은 침입 탐지 성능을 보였음을 확인했음.

**Arilangga (2024) 실험 성능 요약표**

모델 | 정확도 (Accuracy) | 정밀도 (Precision) | 재현율 (Recall) | F1-점수 (F1-Score)
-- | -- | -- | -- | --
Gradient Boosting | 1.00 | 1.00 | 1.00 | 1.00
Random Forest | 1.00 | 1.00 | 1.00 | 1.00
MLP (Multi-Layer Perceptron) | 0.99 | 0.99 | 0.99 | 0.99
Logistic Regression | 0.96 | 0.96 | 0.96 | 0.96

Arilangga에 따르면,
1. Gradient Boosting과 Random Forest가 데이터셋 내 보안 사고를 분류하고 예측하는 데 뛰어난 능력을 가지고 있음을 시사 (Logistic Regression의 경우 0.96으로 더 낮음)
2. 과적합(Overfitting) 가능성: 데이터의 특성상 RT-IoT2022 데이터셋이 실제 IoT 환경을 잘 반영하고 다양한 공격 시나리오를 포함하고 있지만 통제된 환경에서 수집된 데이터임. 모델이 학습 데이터에 과도하게 적합되어 실제 환경에서는 네트워크 트래픽은 예측 불가능한 변동성과 새로운 유형의 공격에 훨씬 더 많이 노출될 수 있음 (즉 일반화 성능이 떨어질 수 있음) 새로운, 보지 못한 데이터에 대한 예측력을 저해할 가능성 존재

하기 때문에 성능이 높게 나온다고 분석함.

## 7 : 향후 개선 및 추가적인 실험 제안

1. 과적합 검증
  - 교차 검증 (Cross-Validation) 강화
  - 독립적인 검증 데이터셋 활용: RT-IoT2022 데이터셋 외에, 가능하다면 다른 실제 IoT 환경에서 수집된 독립적인 데이터셋을 사용하여 모델의 일반화 능력을 교차 검증하는 것이 중요 
  - 학습 곡선 (Learning Curve) 분석: 학습 데이터 크기에 따른 모델 성능 변화를 시각화하여 과적합 또는 과소적합 여부를 진단할 수 있음

2. 경량화
  - 논문에서도 언급했듯, 실시간(Real Time) IoT data를 수집하고 모델을 이용하기 위해서는 정확도 뿐만 아니라 다른 지표들 또한 준수해야 함:
    - 일반화 성능, 낮은 컴퓨팅/전력 자원(Computational Constraints), 확장성 등이 고려돼야 함. 
    - 따라서 시간이 많이 걸리는 고성능 알고리즘 대신 보다 단순하고 빠르게 처리할 수 있는 다른 알고리즘(경량화 트리, 선형 모델 등)들을 통해 실험할 필요가 있음.
